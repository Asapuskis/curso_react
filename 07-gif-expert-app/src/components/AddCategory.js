import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ( {setCategories} ) => {

    // Creamos un useState para almacenar el valor de la caja de texto y la iniciamos a Kps
    const [inputValue, setInputValue] = useState('Kps');

    // Manejar los cambios de la caja de texto mediante useState.
    const handleInputChange = (e) => {
        setInputValue(e.target);
        console.log('handleInputChange llamado');
    }

    // Manejar el submit del formulario.
    const handleSubmit = (e) => {
        // Mediante preventDefault, evitamos que se refresque toda la página y perder los estados.
        e.preventDefault();

        console.log('handleSubmit', inputValue);

        if (inputValue.trim().length > 2) {
            setCategories( cats => [ inputValue, ...cats,]);
            setInputValue('');
            console.log('Cambios hechos');
        }
    }

    // Devolvemos un formulario para añadir la categoría a la lista.
    return (
        <form onSubmit={handleSubmit}>
            <h2>AddCategory</h2>
            <p>{inputValue}</p>
            <input
                type='text'
                value={inputValue}
                onChange={ handleInputChange }
            />
        </form>
    )
}

// Añadimos el propType requerido para la función setCategories
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
