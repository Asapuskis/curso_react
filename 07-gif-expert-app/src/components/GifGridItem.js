import React from 'react'
import PropTypes from 'prop-types';

// Acordarse de desestructurar image ya que por defecto viene como si fuese una prop, entonces o ponemos prop.image o {image}
export const GifGridItem = ( {image} ) => {
  console.log(`title: ${image.title}`)
  console.log(`url: ${image.url}`)
  return (
    <div className='card animate__animated animate__fadeIn '>
        <p>{image.title}</p>
        <img src={image.url} alt={image.title} />
    </div>
  )
}

GifGridItem.propTypes = {
  image: PropTypes.object.isRequired
}
