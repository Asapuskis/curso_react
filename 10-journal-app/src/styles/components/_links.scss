.link {
    text-decoration: none;
    color: $dark-grey;
    margin-bottom: 10px;
}

.link:hover {
    text-decoration: underline;
}