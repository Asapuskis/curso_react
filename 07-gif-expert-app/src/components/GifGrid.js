import React from 'react'
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ( {category}) => {


    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    // Código para mostrar los gifs utilizando los hooks useState y useEffect.

    // const [images, setImages] = useState([]);

    // // Hook useEffect. Este hook sirve para ejecutar cierto código de manera condicional.
    // // De esta manera evitaremos llamar al código múltiples veces cuando no es necesario recargar los componentes o el navegador.
    // // useEffect tiene dos argumentos:
    // // 1. El código que se va a ejecutar, en este caso getGifs()
    // // 2. Una lista de dependencias que se tienen que cumplir para que se ejecute este código. Si ponemos vacío [] se ejecutará una única vez.
    // // Si colocamos category, se llamará a este hook cuando la variable category se modifique.
    // useEffect( () => {
    //         getGifs(category)
    //             .then( imgs => setImages( imgs ));
    //     },
    //     [ category ]
    // );

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    // Código para mostrar los gifs con un custom Hook creado expresamente para esto:
    // Lo ideal es utilizar custom hooks que llaman a hooks ya existentes y colocarlos aparte en el código, se recomienda crear carpeta hooks

    const {data, loading} = useFetchGifs(category);
    console.log(data);
    console.log(loading);

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    return (
        <>
            <h3 className='animate__animated animate__fadeIn'>{category}</h3>
            {/* Mediante un operador terniario, mostramos loading mientras carga los gifs y nada cuando los haya cargado */}
            { loading ? <p className='animate__animated animate__flash'>loading</p> : null}
            <div className='card-grid animate__animated animate__fadeIn'>
                {
                    data.map( img => (
                        <GifGridItem
                            image={img}
                            key={img.id}
                        />
                    ))
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}