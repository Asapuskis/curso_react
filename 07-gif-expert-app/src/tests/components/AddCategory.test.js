import React from 'react';
import { shallow } from 'enzyme';
import { AddCategory } from "../../components/AddCategory";


describe('Pruebas en <AddCategory />', () => {

    const setCategories = jest.fn();
    let wrapper = shallow( <AddCategory setCategories={setCategories} />);

    // beforeEach se ejecuta antes de ejecutar cada test, está bien para cuando se haya cambiado algún valor de un elemento que se va a utilizar
    // en más de un test declarado como global.
    beforeEach( () => {
        jest.clearAllMocks();
        wrapper = shallow( <AddCategory setCategories={setCategories} />);
    })

    test('Debe de mostrarse correctamente el componente', () => {
        expect( wrapper ).toMatchSnapshot();
     })

    test('Debe cambiar la caja de texto', () => {
        const input = wrapper.find('input');
        // change necesita recibir como parámetro el evento e, para definirlo en el simulate, se crea un objeto {} simulando la respuesta del evento.
        // En este caso, el evento tiene una propiedad llamada target con un valor, eso es lo que vamos a simular.
        const value = 'Kptracer';
        const simEvent = {
            target: value
        }
        input.simulate('change', simEvent);
    })

    test('No debe postear la información con Submit', () => {
        // Como se llama al evento preventDefault, creamos una dummy function para simularla.
        const preventDefault = () =>{};
        wrapper.find('form').simulate('submit', {preventDefault})
        expect(setCategories).toHaveBeenCalled();
     })

    test('Debe llamar el setCategories y limpiar la caja de texto', () => {
        const input = wrapper.find('input');
        const value = 'Kptracer';
        const simEvent = {
            target: value
        }
        input.simulate('change', simEvent);
        const preventDefault = () =>{};
        wrapper.find('form').simulate('submit', {preventDefault})
        expect(setCategories).toHaveBeenCalled();
        expect( wrapper.find('input').prop('value')).toBe('');
    })
})