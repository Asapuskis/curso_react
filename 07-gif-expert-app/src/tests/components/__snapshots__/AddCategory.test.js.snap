// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> Debe de mostrarse correctamente el componente 1`] = `
<form
  onSubmit={[Function]}
>
  <h2>
    AddCategory
  </h2>
  <p>
    Kps
  </p>
  <input
    onChange={[Function]}
    type="text"
    value="Kps"
  />
</form>
`;
